packages/react-dom/src/client/ReactDOMFiberInput.js

```diff
@@ -235,32 +213,20 @@ export function updateWrapper(element: Element, props: Object) {
 
 export function postMountWrapper(element: Element, props: Object) {
   var node = ((element: any): InputWithWrapperState);
+  var hasUserInput = node.value !== '';
+  var value = node._wrapperState.initialValue;
 
-  // Detach value from defaultValue. We won't do anything if we're working on
-  // submit or reset inputs as those values & defaultValues are linked. They
-  // are not resetable nodes so this operation doesn't matter and actually
-  // removes browser-default values (eg "Submit Query") when no value is
-  // provided.
+  if (value !== '' || props.hasOwnProperty('value')) {
```

Comment:
`hasUserInput`?

Reply:
Ah this is confusing. `node.value` is different from `node._wrapperState.initialValue`.`hasUserInput` tracks if a user has started typing into an input before JS evaluates (in the case of server-side rendering).

When a component mounts, we don't want to initial set `value` or `defaultValue` if they aren't provided in props, and we don't want to set `value` to the initial React-provide value if a user has already entered text.

I'm going to rename the `value` variable to `initialValue`. I think I can also change this check to avoid type comparison and just check for presence of the `value` or `defaultValue` in `props`.

Reply:
Done in 90ed488